proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,692 @@
+While the double backslash is still required, I think the overall style is much better (could be resolved in a later proposal adding raw strings).
+
+Example from Fourmolu (printf)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+(`link <https://github.com/fourmolu/fourmolu/blob/0b228e12872be8f8e97daf24e82632321fff947f/config/Generate.hs#L146-L165>`_)
+
+With ``unlines``:
+
+::
+
+  unlines
+    [ printf "instance Aeson.FromJSON %s where" fieldTypeName,
+      printf "  parseJSON =",
+      printf "    Aeson.withText \"%s\" $ \\s ->" fieldTypeName,
+      printf "      either Aeson.parseFail pure $",
+      printf "        parsePrinterOptType (Text.unpack s)",
+      printf "",
+      printf "instance PrinterOptsFieldType %s where" fieldTypeName,
+      printf "  parsePrinterOptType s =",
+      printf "    case s of",
+      unlines_
+        [ printf "      \"%s\" -> Right %s" val con
+        | (con, val) <- enumOptions
+        ],
+      printf "      _ ->",
+      printf "        Left . unlines $",
+      printf "          [ \"unknown value: \" <> show s",
+      printf "          , \"Valid values are: %s\"" (renderEnumOptions enumOptions),
+      printf "          ]",
+      printf ""
+    ]
+
+With ``string-interpolate`` (without interpolation, for a fair comparison):
+
+::
+
+  printf
+    [__i|
+    instance Aeson.FromJSON %s where
+      parseJSON =
+        Aeson.withText "%s" $ \s ->
+          either Aeson.parseFail pure $
+            parsePrinterOptType (Text.unpack s)
+
+    instance PrinterOptsFieldType %s where
+      parsePrinterOptType s =
+        case s of
+    %s
+          _ ->
+            Left . unlines $
+              [ "unknown value: " <> show s
+              , "Valid values are: %s"
+              ]
+    |]
+    fieldTypeName
+    fieldTypeName
+    fieldTypeName
+    ( unlines_
+        [ printf "      \"%s\" -> Right %s" val con
+        | (con, val) <- enumOptions
+        ]
+    )
+    (renderEnumOptions enumOptions)
+
+With multiline strings:
+
+::
+
+  printf
+    """
+    instance Aeson.FromJSON %s where
+      parseJSON =
+        Aeson.withText "%s" $ \\s ->
+          either Aeson.parseFail pure $
+            parsePrinterOptType (Text.unpack s)
+
+    instance PrinterOptsFieldType %s where
+      parsePrinterOptType s =
+        case s of
+    %s
+          _ ->
+            Left . unlines $
+              [ "unknown value: " <> show s
+              , "Valid values are: %s"
+              ]
+    """
+    fieldTypeName
+    fieldTypeName
+    fieldTypeName
+    ( unlines_
+        [ printf "      \"%s\" -> Right %s" val con
+        | (con, val) <- enumOptions
+        ]
+    )
+    (renderEnumOptions enumOptions)
+
+Effect and Interactions
+-----------------------
+
+A multiline string should be the same as a normal string after parsing, so ``OverloadedStrings`` and any other language features should work as usual.
+
+Should not break existing code, unless someone is actually using ``"""a"""`` to mean ``"" "a" ""``.
```

Comment:
To clarify the backwards compatibility point: is it possible to have the lexer change behaviour depending on whether `MultilineStrings` is enabled? If so, there's clearly no backwards compatibility concern. But if `"""` will always need to be lexed as a string delimiter (and the multiline construct rejected later if `MultilineStrings` is disabled), we might indeed break code like `f """a"""` where `f` expects three string arguments.

I agree this is pretty unlikely though, and a quick Hackage search suggests there are far more occurrences of packages that use `"""` as a multiline string delimiter (e.g. in packages that include some Python code or TOML). So it's probably fine.

Reply:
As an implementation point: yes it's possible to condition the lexer on the presence of the extension. (This is not possible in the parser, but it works fine in the lexer.)